fix: Disable all emails per event type w/o a workflow [CAL-4066]

joeauyeung opened the pull request:
## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

This PR enables team event types under an org to disable the standard emails without having a workflow set.

There was also some confusion with some users where they believed that the current option would disable all emails. This PR expands disabling emails for the following:
- Scheduled
- Requested bookings
- Request reminders
- Requested denied
- Reschedule
- Requested reschedule
- Cancelled
- Payment requested
- Round robin emails to team members
- Seats specific emails

Before we would only check if we should send the email before calling the email functions. Now we send the whole event type metadata to the email function and have the check logic in the function.

This also introduces a new confirmation if the user wants to disable emails
![CleanShot 2024-07-24 at 23 09 47@2x](https://github.com/user-attachments/assets/ec307057-d3f0-4408-88cb-a0a6cb1255c0)


- Fixes #12920
- Fixes CAL-4066 (Linear issue number - should be visible at the bottom of the GitHub issue description)

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Under an event type in a team in an org you should be able to see the two options under the advanced tab
- When enabling the options a dialog should appear where you have to type confirm to enable the option
- Enable both options
- Under each type of email that can be disabled, create a booking to test that the emails ARE NOT SENT
- Disable the option and ensure emails ARE SENT

joeauyeung committed:
Create confirmation dialog for attendees

joeauyeung committed:
Add disabling host emails option

joeauyeung committed:
Fix DOM issues

joeauyeung committed:
Allow updating disabling emails on backend

joeauyeung committed:
Pass metadata to sendEmails functions

joeauyeung committed:
Clean up

joeauyeung reviewed (commented):
I would look at `email-manager.ts` to see which emails are being affected by this PR.

When getting the metadata there is a lot of optional chaining when getting the event type metadata. This is because dynamic bookings might not have an event type associated with them in the DB. 

I have omitted optional chaining where an event type is guaranteed such as bookings that require confirmation and payments associated with them.

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

joeauyeung committed:
Type fixes

joeauyeung committed:
Fix test

CarinaWolli reviewed (commented):
Early review: 

- We said to allow disabling all emails without workflows only for orgs. But now every user can disable ALL emails only by having one workflow that triggers when booking is created.  
  - we either allow it for all users to disable emails without workflows 
  - or, non-org users can only disable the booking confirmation email (not all emails) when the workflows exists --> same as it was before 
- What happens to event types that already had the confirmation email disabled before? To me it looks like this would now change for them and all emails would stop sending
- Orgs can only disable the emails for team event types. I think we want that for personal event types as well

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

joeauyeung committed:
Allow for enterprise only

joeauyeung commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -569,6 +573,34 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
           )}
         />
       )}
+      {hasValidLicense && (
```
We now only allow these options for enterprise customers. We'll still keep both options available but will replace once we allow users to set granular options for emails
![CleanShot 2024-07-25 at 15 18 50@2x](https://github.com/user-attachments/assets/e2d565d8-5a89-45d1-8140-8debdee3132f)

joeauyeung commented on packages/prisma/zod-utils.ts:
```diff
@@ -81,6 +81,12 @@ export const EventTypeMetaDataSchema = z
     disableSuccessPage: z.boolean().optional(),
     disableStandardEmails: z
       .object({
+        all: z
+          .object({
+            host: z.boolean().optional(),
+            attendee: z.boolean().optional(),
+          })
+          .optional(),
```
We now have a new option to disable all emails while still keeping the old option for just confirmation emails.

joeauyeung reviewed (commented):
@CarinaWolli addressed your feedback.

I don't think there would be a lot of use cases where an individual wants to disable all emails as it takes a lot of setup which might be specific to their company. If a company did want this for individual events they should use the managed event type.

joeauyeung committed:
Type fix

joeauyeung committed:
Test fix

joeauyeung committed:
Check for parentId

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

CarinaWolli commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -569,6 +573,34 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
+        />
+      )}
+      {team?.parentId && (
+        <Controller
+          name="metadata.disableStandardEmails.all.host"
+          defaultValue={!!formMethods.getValues("seatsPerTimeSlot")}
+          render={({ field: { value, onChange } }) => (
+            <>
+              <DisableAllEmailsSetting checked={value} onCheckedChange={onChange} recipient="hosts" t={t} />
+            </>
+          )}
+        />
```
I think we can move this up right after themetadata.disableStandardEmails.all.attendee Controller. No need check for `team?.parentId` again

CarinaWolli commented on apps/web/components/eventtype/EventAdvancedTab.tsx:
```diff
@@ -56,6 +58,8 @@ export const EventAdvancedTab = ({ eventType, team }: Pick<EventTypeSetupProps,
   );
   const [hashedUrl, setHashedUrl] = useState(eventType.hashedLink?.link);
   const bookingFields: Prisma.JsonObject = {};
+  const session = useSession();
+  const hasValidLicense = session.data ? session.data.hasValidLicense : null;
```
`hasValidLicense` isn't used

CarinaWolli commented on apps/web/components/eventtype/settings/DisableAllEmailsSetting.tsx:
```diff
@@ -0,0 +1,81 @@
+          <p className="text-default text-sm">
+            <Trans i18nKey="disable_attendees_emails_description">
+              This will disable all emails to {{ recipient }}. This includes booking confirmations, requests,
+              reschedules and reschedule requests, and cancellation emails.
```
We are missing the location change emails here (maybe some other email too)

Should we add something like 'and any other emails related to booking updates'

CarinaWolli commented on apps/web/components/eventtype/settings/DisableAllEmailsSetting.tsx:
```diff
@@ -0,0 +1,81 @@
+              <br />
+              <br />
+              It is your responsibility to ensure that your {{ recipient }} are aware of any bookings and
+              changes to their booking.
```
```suggestion
              changes to their bookings.
```

CarinaWolli reviewed (changes requested):
- [ ]  `sendScheduledEmails` and `sendScheduledSeatsEmails` doesn't support disabling all emails (only the old feature to disable confirmation email when workflow enabled is supported)
- [ ] In `eventTypes/update.handler.ts` we need to check if the event type is from an org before updating `metadata.disableStandardEmails.all`
- [ ] we also should not send reminder emails in`scheduleMandatoryReminder` if all emails to attendees are disabled

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

joeauyeung committed:
Disable emails for sendScheduledEmails and sendScheduledSeatsEmails

joeauyeung committed:
Remove unused variable

joeauyeung committed:
Only update disable all emails if event type belongs to an org

joeauyeung committed:
Disable scheduling mandatory reminder workflow if attendee emails are disabled

joeauyeung committed:
Refactor disable email UI

joeauyeung committed:
Update copy

joeauyeung commented:
> * [ ]   `sendScheduledEmails` and `sendScheduledSeatsEmails` doesn't support disabling all emails (only the old feature to disable confirmation email when workflow enabled is supported)
> * [ ]  In `eventTypes/update.handler.ts` we need to check if the event type is from an org before updating `metadata.disableStandardEmails.all`
> * [ ]  we also should not send reminder emails in`scheduleMandatoryReminder` if all emails to attendees are disabled

Addressed all feedback

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

CarinaWolli commented on packages/trpc/server/routers/viewer/eventTypes/update.handler.ts:
```diff
@@ -255,9 +259,15 @@ export const updateHandler = async ({ ctx, input }: UpdateOptions) => {
-  if (input.metadata?.disableStandardEmails) {
-    //check if user is allowed to disabled standard emails
+  if (input.metadata?.disableStandardEmails?.all) {
+    if (!eventType.team.parentId) {
```
type error here `'eventType.team' is possibly 'null'`

CarinaWolli commented on packages/emails/email-manager.ts:
```diff
@@ -201,7 +242,7 @@ export const sendScheduledSeatsEmails = async (
   const emailsToSend: Promise<unknown>[] = [];
 
-  if (!hostEmailDisabled) {
+  if (!hostEmailDisabled && !eventTypeDisableHostEmail(eventTypeMetadata)) {
```
`eventTypeMetadata` is missing as a param

CarinaWolli reviewed (changes requested):
Found some places where we don't pass `eventTypeMetadata` when calling the email function, let me know in case it's not needed there:
- [ ] `sendRoundRobinCancelledEmails` in new `roundRobinReassignment` function 
- [ ] `sendCancelledEmails` in `handleDeleteCredential`
- [ ] `sendRequestRescheduleEmail` in `app-store/vital/lib/reschedule.ts`

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

joeauyeung committed:
Type fixes

joeauyeung committed:
Merge branch 'disable-attendee-emails' of https://github.com/calcom/cal.com into disable-attendee-emails

joeauyeung committed:
Merge branch 'main' into disable-attendee-emails

joeauyeung committed:
Fix missing params

CarinaWolli reviewed (approved):
All my comments have been addressed, looks good now 🙏

YPC-Avishai commented:
Is this feature available for teams on cal.com, too? 
Because I can not find those toggles. @joeauyeung

janrabek commented:
I can confirm @YPC-Avishai

We really need this feature, does anyone know when it's going to be released on cal.com?

nakul-malhotra commented:
is this ever going to be introduced for individuals? why are we blocking this to enterprise only?
